File: supabase/functions/_shared/checkout_origin.ts
// Which surface a Stripe Checkout session was started from, recorded in the
// session's metadata as `origin` (macOS Phase 9).
//
// The Mac app is a direct download with no storefront behind it, so macOS sales
// show up in no App Store or Play report — Stripe metadata is the only place
// they can be told apart from phone sales.
//
// NOT the HTTP Origin header — that is `_shared/cors.ts`. This is a claim in
// the request body, so it is validated against the list rather than echoed: a
// tampered client gets `unknown`, never arbitrary metadata. Keep in step with
// lib/services/checkout_origin.dart.

const ALLOWED = new Set(["ios", "android", "macos", "web"]);

export function checkoutOrigin(raw: unknown, fallback = "unknown"): string {
  return typeof raw === "string" && ALLOWED.has(raw) ? raw : fallback;
}

File: lib/services/checkout_origin.dart
import 'package:flutter/foundation.dart';

/// Which surface a Stripe Checkout session was started from.
///
/// Sent with every checkout request and recorded in the session's Stripe
/// metadata as `origin`. The Mac app is a direct download with no storefront
/// behind it, so macOS sales appear in no App Store or Play report — this tag
/// is the only place they can be told apart from phone sales.
///
/// The edge functions validate against the same set rather than echoing what
/// the client sends, so a tampered build writes `unknown`, not arbitrary
/// metadata. Keep this list and `_shared/checkout_origin.ts` in step.
class NileCheckoutOrigin {
  NileCheckoutOrigin._();

  static String get current {
    if (kIsWeb) return 'web';
    return switch (defaultTargetPlatform) {
      TargetPlatform.iOS => 'ios',
      TargetPlatform.android => 'android',
      TargetPlatform.macOS => 'macos',
      _ => 'unknown',
    };
  }
}
